extensions/amp-story/1.0/amp-story.js

```diff
@@ -485,6 +485,8 @@ export class AmpStory extends AMP.BaseElement {
       Action.TOGGLE_PAUSED,
       this.pausedStateToRestore_
     );
+    // Simulate touchend to avoid touchstarts to not get resolved when swiping between stories(#28425)
+    this.element.dispatchEvent(createCustomEvent(this.win, 'touchend', null));
```

Comment:
This works but makes things a bit unpredictable.
What I'm worried about off the top of my head:
- Losing focus on some element
-  Sending touchend will switch the `paused` state back to `false` from `page-advancement`
- If we switch from navigation on `click` to navigation on `touchend` this would navigate and be hard to debug (we might do that in a near future because iOS is extremely slow at emulating `click` from a `touchend`)

What do you think about replicating the logic that triggers this `resume_` (visibilitychange events) into `page-advancement.ManualAdvancement`, and run a few lines from the `ManualAdvancement.onTouchend_()` method to reset what needs to be?

I think these are the lines we care about:
https://github.com/ampproject/amphtml/blob/769e8152bc6897844b8778a6d447fe99e5015c7c/extensions/amp-story/1.0/page-advancement.js#L370-L379

Reply:
Makes sense. I wanted to do it this way but all the methods on the ManualAdvancement were private so I assumed the class was abstracted from the story. However, instantiated the amp-story.advancement_ field as a ManualAdvancement which makes sense, and I think that makes this approach better.

The paused state change is also necessary since the main way of pausing is touchstart on mobile (where this function gets called), and the test reflects it.

Added those changes, let me know how this looks!

Reply:
Your approach looks good, but overall we've been trying to keep features contained into their class, to make amp-story.js smaller with less interactions with other components. This will scale far better and let us do many improvements.

Can we:
1. move the `this.getAmpDoc().onVisibilityChanged(() => this.onVisibilityChanged_());` into page-advancement and call the method directly from there? So that'd be no change to `amp-story.js`
2. Fix the `PAUSED_STATE`: here you're playing the story no matter what. The story is sometimes paused (user consent, expanded tweet, link tooltip visible...) and now swiping back to the story would force play it no matter what